File: src/com/dutch_computer_technology/mySurveillance/pages/Cameras.java
package com.dutch_computer_technology.mySurveillance.pages;

import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import javax.imageio.ImageIO;

import com.dutch_computer_technology.mySurveillance.Main;
import com.dutch_computer_technology.mySurveillance.Main.printType;
import com.dutch_computer_technology.mySurveillance.accounts.User;
import com.dutch_computer_technology.mySurveillance.cameras.Camera;
import com.dutch_computer_technology.mySurveillance.cameras.Cameras.CameraType;
import com.dutch_computer_technology.mySurveillance.cameras.StreamManager.StreamType;
import com.dutch_computer_technology.mySurveillance.cameras.paths.Path;
import com.dutch_computer_technology.mySurveillance.cameras.reasons.Reason;
import com.dutch_computer_technology.mySurveillance.cameras.streams.Stream;
import com.dutch_computer_technology.mySurveillance.exceptions.JSONParse;
import com.dutch_computer_technology.mySurveillance.json.JSONObject;
import com.dutch_computer_technology.mySurveillance.website.data.Page;
import com.dutch_computer_technology.mySurveillance.website.data.Request;
import com.dutch_computer_technology.mySurveillance.main.ByteManager;
import com.dutch_computer_technology.mySurveillance.main.FileManager.PathType;
import com.dutch_computer_technology.mySurveillance.main.FileManager.ReasonType;
import com.dutch_computer_technology.mySurveillance.main.MySurveillance;
import com.dutch_computer_technology.mySurveillance.main.Random;
import com.dutch_computer_technology.mySurveillance.website.ContentType;
import com.dutch_computer_technology.mySurveillance.website.PageHandler;
import com.dutch_computer_technology.mySurveillance.website.Website.FileType;

public class Cameras implements PageHandler {
	
	private String getDate(LocalDateTime now) {
		
		//Date
		String year = Integer.toString(now.getYear());
		String month = Integer.toString(now.getMonthValue());
		if (month.length() < 2) month = "0" + month;
		String day = Integer.toString(now.getDayOfMonth());
		if (day.length() < 2) day = "0" + day;
		//Time
		String hour = Integer.toString(now.getHour());
		if (hour.length() < 2) hour = "0" + hour;
		String min = Integer.toString(now.getMinute());
		if (min.length() < 2) min = "0" + min;
		String sec = Integer.toString(now.getSecond());
		if (sec.length() < 2) sec = "0" + sec;
		
		return year + "/" + month + "/" + day + " " + hour + ":" + min + ":" + sec;
		
	};
	
	private final int width = 320;
	private final int height = 176;
	
	private byte[] getDefaultStream(MySurveillance ms) {
		
		LocalDateTime now = LocalDateTime.now();
		BufferedImage defImage = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
		Graphics2D g = defImage.createGraphics();
		g.setFont(new Font("Serif", Font.PLAIN, 11));
		g.setColor(Color.white);
		g.drawString(getDate(now), 3, 11);
		g.setFont(new Font("Serif", Font.PLAIN, 32));
		if (now.getSecond() % 2 == 0) g.setColor(Color.red);
		
		String text = ms.lang.get("cameras.noCam");
		int x = (int) (text.length() * 8.75);
		g.drawString(text, (width/2)-x, (height/2)+6);
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		try {
			ImageIO.write(defImage, "jpg", baos);
		} catch (IOException e) {
			Main.print(this, "Could not create default stream", e);
		};
		return baos.toByteArray();
		
	};
	
	public Page get(MySurveillance ms, Request request, User user) {
		
		Map<String, String> query = request.query();
		if (query.containsKey("camera")) {
			
			String name = query.get("camera");
			Camera cam = ms.cameras.getCamera(name);
			
			Page page = new Page();
			byte[] s = new byte[0];
			if (cam == null) {
				s = ByteManager.append(s, "HTTP/1.1 404 Not Found\r\nServer: MySurveillance");
				s = ByteManager.append(s, "\r\nContent-Type: " + ContentType.JPEG.toString());
				s = ByteManager.append(s, "\r\nConnection: Close\r\nCache-Control: no-cache, no-store\r\n\r\n");
				s = ByteManager.append(s, getDefaultStream(ms));
			} else {
				s = ByteManager.append(s, "HTTP/1.1 200 OK\r\nServer: MySurveillance");
				s = ByteManager.append(s, "\r\nContent-Type: " + ContentType.JPEG.toString());
				s = ByteManager.append(s, "\r\nConnection: Close\r\nCache-Control: no-cache, no-store\r\n\r\n");
				byte[] stream = cam.getStreamData();
				if (stream == null || !cam.isRunning()) {
					s = ByteManager.append(s, getDefaultStream(ms));
				} else {
					if (cam.isGlitch()) {
						Random ran = new Random();
						if (ran.action(20)) {
							int index = ran.randomInt(stream.length);
							int size = ran.randomInt(1, 4096);
							byte[] a = ByteManager.between(stream, 0, index);
							byte[] b = ByteManager.between(stream, index+size, stream.length);
							stream = new byte[0];
							Random r = new Random();
							stream = ByteManager.append(stream, a);
							stream = ByteManager.append(stream, r.id(size).getBytes());
							stream = ByteManager.append(stream, b);
						};
					};
					s = ByteManager.append(s, stream);
				};
			};
			page.setRawData(s);
			return page;
			
		};
		
		if (!user.isAdmin()) return ms.website.unauthorizedPage(ContentType.HTML, ms.lang.get("unauthorized.privileges"));
		
		Page page = new Page();
		page.setStatus(200);
		page.setContentType(ContentType.HTML);
		
		String html = new String(ms.website.loadFile(FileType.HTML, "cameras"));
		
		String screens = "<div class=\"screens\">";
		int min = ms.screens.size();
		double sqr = Math.sqrt(min);
		while (sqr % 1 != 0) {
			min++;
			ms.screens.add("");
			sqr = Math.sqrt(min);
		};
		double pro = 100 / sqr;
		int font = 22;
		String size = "calc(" + Double.toString(pro) + "% - 2px);";
		for (int i = 0; i < ms.screens.size() || i < min; i++) {
			String name = "";
			if (i < ms.screens.size()) {
				name = ms.screens.get(i);
			};
			screens += "<div style=\"width: " + size + "; height: " + size + ";\" class=\"screen\"><div class=\"button\" onclick=\"setScreen(event, " + i + ", false);\"></div><h3 class=\"title\" style=\"font-size: " + font + "px;\">" + name + "</h3></div>";
		};
		screens += "</div>";
		html = ms.website.replaceHTML(html, "<min>", Integer.toString(min));
		html = ms.website.replaceHTML(html, "<screens>", screens);
		
		String camNames = "\"\"";
		String camerasJSON = "";
		String cams = "<div class=\"cameras\"><div class=\"newCamera\"><button onclick=\"openSettings(false);\"><img src=\"new.png\"/></button></div><div class=\"divider\"></div>";
		List<String> names = new ArrayList<String>(ms.cameras.getCameras());
		for (String name : names) {
			Camera cam = ms.cameras.getCamera(name);
			if (cam == null) continue;
			camNames += ",\"" + name + "\"";
			cams += "<div class=\"camera\">"
					+ "<div class=\"view\"><canvas src=\"/cameras?camera=" + name + "\"></canvas></div>"
					+ "<div class=\"info\"><p>Name: " + name + "</p><p>Status: " + (cam.isOnline() ? (cam.isGlitch() ? "Glitch" : "Online") : "Offline") + "</p></div>"
					+ "<div class=\"options\"><button onclick=\"openSettings('" + name + "');\"><img src=\"settings.png\"/></button>"
					+ "<button onclick=\"power('" + name + "');\"><img src=\"power.png\"/></button>"
					+ "<button onclick=\"glitch('" + name + "');\"><img src=\"glitch.png\"/></button></div>"
					+ "</div>";
			cams += "<div class=\"divider\"></div>";
			camerasJSON += "\"" + cam.getName() + "\":" + cam.toString() + ",";
		};
		camerasJSON = camerasJSON.substring(0, camerasJSON.length() - 1);
		cams = cams.substring(0, cams.length() - 27);
		cams += "</div>";
		html = ms.website.replaceHTML(html, "<cameras>", cams);
		
		html = ms.website.loadHTML(html);
		html = ms.website.replaceHTML(html, "<camerasJSON>", camerasJSON);
		html = ms.website.replaceHTML(html, "<camNames>", camNames);
		
		String cameraTypes = "";
		for (CameraType type : CameraType.values()) {
			cameraTypes += "\"" + type.toString() + "\",";
		};
		cameraTypes = cameraTypes.substring(0, cameraTypes.length() - 1);
		
		String streamTypes = "";
		for (StreamType type : StreamType.values()) {
			streamTypes += "\"" + type.toString() + "\",";
		};
		streamTypes = streamTypes.substring(0, streamTypes.length() - 1);
		
		String pathTypes = "";
		for (PathType type : PathType.values()) {
			pathTypes += "\"" + type.toString() + "\",";
		};
		pathTypes = pathTypes.substring(0, pathTypes.length() - 1);
		
		String reasonTypes = "";
		for (ReasonType type : ReasonType.values()) {
			reasonTypes += "\"" + type.toString() + "\",";
		};
		reasonTypes = reasonTypes.substring(0, reasonTypes.length() - 1);
		
		html = ms.website.replaceHTML(html, "<cameraTypes>", cameraTypes);
		html = ms.website.replaceHTML(html, "<streamTypes>", streamTypes);
		html = ms.website.replaceHTML(html, "<pathTypes>", pathTypes);
		html = ms.website.replaceHTML(html, "<reasonTypes>", reasonTypes);
		
		page.setData(html);
		return page;
		
	};
	
	public Page post(MySurveillance ms, Request request, User user) {
		
		if (!user.isAdmin()) return ms.website.unauthorizedPage(ContentType.TEXT, ms.lang.get("unauthorized.privileges"));
		
		try {
			
			JSONObject json = new JSONObject(request.data());
			
			String cmd = json.getString("cmd");
			if (cmd == null) {
				Page page = new Page();
				page.setStatus(400);
				page.setData(ms.lang.get("data.incomplete"));
				return page;
			};
			
			Main.print(this, printType.INFO, "Admin " + user.getUsername() + " on IP: " + request.ip() + ", used the following command: " + cmd);
			
			if (cmd.equals("getStream")) {
				
				String typeS = json.getString("type");
				StreamType type = StreamType.Disabled;
				if (typeS != null) {
					try {
						type = StreamType.valueOf(typeS);
					} catch(IllegalArgumentException ignore) {};
				};
				
				if (type == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.streamTypeNotFound"));
					return page;
				};
				
				String name = json.getString("name");
				Camera cam = null;
				if (name != null) cam = ms.cameras.getCamera(name);
				
				Stream stream = ms.streamManager.getStream(cam, type);
				if (stream == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.streamNotFound"));
					return page;
				};
				
				Page page = new Page();
				page.setStatus(200);
				page.setContentType(ContentType.HTML);
				page.setData(stream.html());
				
				return page;
				
			};
			
			if (cmd.equals("getPath")) {
				
				String typeS = json.getString("type");
				PathType type = PathType.Relative;
				if (typeS != null) {
					try {
						type = PathType.valueOf(typeS);
					} catch(IllegalArgumentException ignore) {};
				};
				
				if (type == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.pathTypeNotFound"));
					return page;
				};
				
				String name = json.getString("name");
				Camera cam = null;
				if (name != null) cam = ms.cameras.getCamera(name);
				
				Path path = ms.fileManager.getPath(cam, type);
				if (path == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.pathNotFound"));
					return page;
				};
				
				Page page = new Page();
				page.setStatus(200);
				page.setContentType(ContentType.HTML);
				page.setData(path.html());
				
				return page;
				
			};
			
			if (cmd.equals("getReason")) {
				
				String typeS = json.getString("type");
				ReasonType type = ReasonType.MaxSize;
				if (typeS != null) {
					try {
						type = ReasonType.valueOf(typeS);
					} catch(IllegalArgumentException ignore) {};
				};
				
				if (type == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.reasonTypeNotFound"));
					return page;
				};
				
				String name = json.getString("name");
				Camera cam = null;
				if (name != null) cam = ms.cameras.getCamera(name);
				
				Reason reason = ms.fileManager.getReason(cam, type);
				if (reason == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.reasonNotFound"));
					return page;
				};
				
				Page page = new Page();
				page.setStatus(200);
				page.setContentType(ContentType.HTML);
				page.setData(reason.html());
				
				return page;
				
			};
			
			if (cmd.equals("power")) {
				
				String name = json.getString("name");
				if (name == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				Camera cam = ms.cameras.getCamera(name);
				if (cam == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.camNotFound"));
					return page;
				};
				
				cam.isOnline(!cam.isOnline());
				ms.cameras.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("glitch")) {
				
				String name = json.getString("name");
				if (name == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				Camera cam = ms.cameras.getCamera(name);
				if (cam == null) {
					Page page = new Page();
					page.setStatus(404);
					page.setData(ms.lang.get("cameras.camNotFound"));
					return page;
				};
				
				cam.isGlitch(!cam.isGlitch());
				ms.cameras.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("min")) {
				
				if (ms.screens.size() < 2) {
					Page page = new Page();
					page.setStatus(403);
					return page;
				};
				
				ms.screens.remove(ms.screens.size() - 1);
				int min = ms.screens.size();
				double sqr = Math.sqrt(min);
				while (sqr % 1 != 0) {
					min--;
					ms.screens.remove(ms.screens.size() - 1);
					sqr = Math.sqrt(min);
				};
				ms.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("plus")) {
				
				ms.screens.add("");
				int min = ms.screens.size();
				double sqr = Math.sqrt(min);
				while (sqr % 1 != 0) {
					min++;
					ms.screens.add("");
					sqr = Math.sqrt(min);
				};
				ms.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("setScreen")) {
				
				if (!json.contains("screen")) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				int screen = json.getInt("screen");
				
				String name = json.getString("name");
				if (name == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				if (screen > ms.screens.size()) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("cameras.screenNotFound"));
					return page;
				};
				
				ms.screens.set(screen, name);
				ms.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("save")) {
				
				JSONObject data = json.getJSONObject("data");
				if (data == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				String name = data.getString("name");
				if (name == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				String oldName = data.getString("oldName");
				if (oldName == null) oldName = name;
				
				Camera cam = ms.cameras.getCamera(oldName);
				if (cam == null) cam = new Camera(ms, oldName);
				
				if (!oldName.equals(name)) {
					if (ms.cameras.getCamera(name) != null) {
						Page page = new Page();
						page.setStatus(400);
						page.setData(ms.lang.get("accounts.nameUsed"));
						return page;
					};
				};
				
				cam.setName(name);
				
				String cameraType = data.getString("cameraType");
				if (cameraType != null) {
					try {
						CameraType cType = CameraType.valueOf(cameraType);
						cam.setType(cType);
					} catch(IllegalArgumentException ignore) {};
				};
				
				JSONObject stream = data.getJSONObject("stream");
				if (stream != null) {
					String type = stream.getString("type");
					if (type != null) {
						StreamType sType = null;
						try {
							sType = StreamType.valueOf(type);
						} catch(IllegalArgumentException ignore) {};
						if (sType != null) {
							cam.setStream(ms.streamManager.createStream(sType, cam, stream));
						};
					};
				};
				
				JSONObject path = data.getJSONObject("path");
				if (path != null) {
					String type = path.getString("type");
					if (type != null) {
						PathType pType = null;
						try {
							pType = PathType.valueOf(type);
						} catch(IllegalArgumentException ignore) {};
						if (pType != null) {
							cam.setPath(ms.fileManager.createPath(pType, cam, path));
						};
					};
				};
				
				JSONObject reason = data.getJSONObject("reason");
				if (reason != null) {
					String type = path.getString("type");
					if (type != null) {
						ReasonType rType = null;
						try {
							rType = ReasonType.valueOf(type);
						} catch(IllegalArgumentException ignore) {};
						if (rType != null) {
							cam.setReason(ms.fileManager.createReason(rType, cam, reason));
						};
					};
				};
				
				ms.cameras.addCamera(cam);
				ms.cameras.save();
				
				return new Page();
				
			};
			
			if (cmd.equals("remove")) {
				
				String name = json.getString("name");
				if (name == null) {
					Page page = new Page();
					page.setStatus(400);
					page.setData(ms.lang.get("data.incomplete"));
					return page;
				};
				
				ms.cameras.remCamera(name);
				ms.cameras.save();
				
				return new Page();
				
			};
			
		} catch (JSONParse e) {
			Main.print(this, "Can't read data", e);
			Page page = new Page();
			page.setStatus(400);
			page.setData(ms.lang.get("data.parse"));
			return page;
		};
		
		return null;
		
	};
	
};

File: src/com/dutch_computer_technology/mySurveillance/pages/js/cameras.js
let camNames = [<camNames>];
let cameraTypes = [<cameraTypes>];
let streamTypes = [<streamTypes>];
let pathTypes = [<pathTypes>];
let reasonTypes = [<reasonTypes>];
let cameras = {<camerasJSON>};

function setScreen(event, screen, name) {
	
	if (name !== false) {
		let xhr = new XMLHttpRequest();
		xhr.open("POST", "/cameras");
		xhr.onreadystatechange = function() {
			if (this.readyState != 4) return;
			if (this.status == 200) {
				location.reload();
			} else {
				let sensor = document.getElementById("sensor");
				if (sensor) sensor.remove();
			};
		};
		let sendData = {"cmd":"setScreen","screen":screen,"name":name};
		xhr.send(JSON.stringify(sendData));
		return;
	};
	
	let sensor = document.getElementById("sensor");
	if (sensor) sensor.remove();
	
	let list = "";
	for (cName of camNames) {
		list += "<option class=\\\"option\\\" value=\\\"" + cName + "\\\">" + cName + "</option>";
	};
	
	let div = document.createElement("div");
	div.id = "sensor";
	div.classList = "sensor";
	div.innerHTML = "<select id=\\\"list\\\" class=\\\"list\\\" style=\\\"top: " + event.y + "px; left: " + event.x + "px;\\\">" + list + "</select>";
	div.addEventListener("click", function(event) {
		let elem = event.target;
		if (elem.id == "sensor") {
			let value = "";
			for (elem of elem.children) {
				if (elem.id != "list") continue;
				value = elem.value;
				break;
			};
			elem.remove();
			setScreen(false, screen, value);
		};
	});
	document.body.appendChild(div);
	
};
function screensMin() {
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status == 200) location.reload();
	};
	let sendData = {"cmd":"min"};
	xhr.send(JSON.stringify(sendData));
};
function screensPlus() {
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status == 200) location.reload();
	};
	let sendData = {"cmd":"plus"};
	xhr.send(JSON.stringify(sendData));
};

function closeSettings() {
	let sensor = document.getElementById("sensor");
	if (sensor) sensor.remove();
};
function saveSettings(name) {
	
	let sensor = document.getElementById("sensor");
	if (!sensor) return;
	
	let data = {};
	
	let nameElem = document.getElementById("name");
	if (name) data["oldName"] = name;
	if (nameElem) {
		data["name"] = nameElem.value;
		console.log(data["name"]);
		if (data["name"].length == 0) {
			inputError(nameElem, "No name");
			return;
		};
	};
	clearInputError(nameElem);
	
	let cameraTypeElem = document.getElementById("cameraType");
	if (cameraTypeElem) data["cameraType"] = cameraTypeElem.value;
	
	let streamTypeElem = document.getElementById("streamType");
	if (streamTypeElem) data["stream"] = {"type":streamTypeElem.value};
	
	let pathTypeElem = document.getElementById("pathType");
	if (pathTypeElem) data["path"] = {"type":pathTypeElem.value};
	
	let reasonTypeElem = document.getElementById("reasonType");
	if (reasonTypeElem) data["reason"] = {"type":reasonTypeElem.value};
	
	if (data.hasOwnProperty("stream")) {
		let stream = document.getElementById("stream");
		if (stream) {
			let streamInputs = stream.getElementsByTagName("input");
			for (streamInput of streamInputs) {
				if (!streamInput.hasAttribute("key")) continue;
				let value = streamInput.value;
				if (streamInput.type == "checkbox") value = streamInput.checked;
				data["stream"][streamInput.getAttribute("key")] = value;
			};
			let streamSelects = stream.getElementsByTagName("select");
			for (streamSelect of streamSelects) {
				if (!streamSelect.hasAttribute("key")) continue;
				data["stream"][streamSelect.getAttribute("key")] = streamSelect.value;
			};
		};
	};
	
	if (data.hasOwnProperty("path")) {
		let path = document.getElementById("path");
		if (path) {
			let pathInputs = path.getElementsByTagName("input");
			for (pathInput of pathInputs) {
				if (!pathInput.hasAttribute("key")) continue;
				let value = pathInput.value;
				if (pathInput.type == "checkbox") value = pathInput.checked;
				data["path"][pathInput.getAttribute("key")] = value;
			};
			let pathSelects = path.getElementsByTagName("select");
			for (pathSelect of pathSelects) {
				if (!pathSelect.hasAttribute("key")) continue;
				data["path"][pathSelect.getAttribute("key")] = pathSelect.value;
			};
		};
	};
	
	if (data.hasOwnProperty("reason")) {
		let reason = document.getElementById("reason");
		if (reason) {
			let reasonInputs = reason.getElementsByTagName("input");
			for (reasonInput of reasonInputs) {
				if (!reasonInput.hasAttribute("key")) continue;
				let value = reasonInput.value;
				if (reasonInput.type == "checkbox") value = reasonInput.checked;
				data["reason"][reasonInput.getAttribute("key")] = value;
			};
			let reasonSelects = reason.getElementsByTagName("select");
			for (reasonSelect of reasonSelects) {
				if (!reasonSelect.hasAttribute("key")) continue;
				data["reason"][reasonSelect.getAttribute("key")] = reasonSelect.value;
			};
		};
	};
	
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status != 200) {
			let buttons = document.getElementById("buttons");
			inputError(buttons, this.response, true);
			return;
		};
		location.reload();
	};
	let sendData = {"cmd":"save","data":data};
	xhr.send(JSON.stringify(sendData));
	
};
async function getStream(data, name) {
	let type = "";
	if (data.hasOwnProperty("stream")) {
		if (data["stream"].hasOwnProperty("type")) {
			type = data["stream"]["type"];
		};
	};
	
	let html = "<div class=\\\"stream\\\"><div class=\\\"options\\\"><p>StreamType:</p><select id=\\\"streamType\\\">";
	for (myType of streamTypes) {
		html += "<option value=\\\"" + myType + "\\\"" + (type == myType ? " selected" : "") + ">" + myType + "</option>";
	};
	html += "</select></div>";
	html += "<div id=\\\"stream\\\" class=\\\"inner\\\">" + await loadStream(type, name) + "</div></div>";
	return html;
};
function loadStream(type, name) {
	return new Promise(function (resolve) {
		let xhr = new XMLHttpRequest();
		xhr.open("POST", "/cameras");
		xhr.onreadystatechange = function() {
			if (this.readyState != 4) return;
			resolve(this.response);
		};
		let sendData = {"cmd":"getStream","type":type,"name":name};
		xhr.send(JSON.stringify(sendData));
	});
};
async function getPath(data, name) {
	let type = "";
	if (data.hasOwnProperty("path")) {
		if (data["path"].hasOwnProperty("type")) {
			type = data["path"]["type"];
		};
	};
	
	let html = "<div class=\\\"path\\\"><div class=\\\"options\\\"><p>PathType:</p><select id=\\\"pathType\\\">";
	for (myType of pathTypes) {
		html += "<option value=\\\"" + myType + "\\\"" + (type == myType ? " selected" : "") + ">" + myType + "</option>";
	};
	html += "</select></div>";
	html += "<div id=\\\"path\\\" class=\\\"inner\\\">" + await loadPath(type, name) + "</div></div>";
	return html;
};
function loadPath(type, name) {
	return new Promise(function (resolve) {
		let xhr = new XMLHttpRequest();
		xhr.open("POST", "/cameras");
		xhr.onreadystatechange = function() {
			if (this.readyState != 4) return;
			resolve(this.response);
		};
		let sendData = {"cmd":"getPath","type":type,"name":name};
		xhr.send(JSON.stringify(sendData));
	});
};
async function getReason(data, name) {
	let type = "";
	if (data.hasOwnProperty("reason")) {
		if (data["reason"].hasOwnProperty("type")) {
			type = data["reason"]["type"];
		};
	};
	
	let html = "<div class=\\\"reason\\\"><div class=\\\"options\\\"><p>ReasonType:</p><select id=\\\"reasonType\\\">";
	for (myType of reasonTypes) {
		html += "<option value=\\\"" + myType + "\\\"" + (type == myType ? " selected" : "") + ">" + myType + "</option>";
	};
	html += "</select></div>";
	html += "<div id=\\\"reason\\\" class=\\\"inner\\\">" + await loadReason(type, name) + "</div></div>";
	return html;
};
function loadReason(type, name) {
	return new Promise(function (resolve) {
		let xhr = new XMLHttpRequest();
		xhr.open("POST", "/cameras");
		xhr.onreadystatechange = function() {
			if (this.readyState != 4) return;
			resolve(this.response);
		};
		let sendData = {"cmd":"getReason","type":type,"name":name};
		xhr.send(JSON.stringify(sendData));
	});
};
async function openSettings(name) {
	
	let sensor = document.getElementById("sensor");
	if (sensor) sensor.remove();
	
	let data = cameras.hasOwnProperty(name) ? cameras[name] : {};
	let _name = name ? name : "";
	let _cameraType = data.hasOwnProperty("cameraType") ? data["cameraType"] : "";
	
	let cameraOptions = "<div><p>CameraType:</p><select id=\\\"cameraType\\\">";
	for (type of cameraTypes) {
		cameraOptions += "<option value=\\\"" + type + "\\\"" + (_cameraType == type ? " selected" : "") + ">" + type + "</option>";
	};
	cameraOptions += "</select></div>";
	
	let rem = "";
	if (name) rem = "<button onclick=\\\"removeCamera('" + name + "');\\\">{{input.remove}}</button>";
	let save = "<button onclick=\\\"saveSettings();\\\">{{input.save}}</button>";
	if (name) save = "<button onclick=\\\"saveSettings('" + name + "');\\\">{{input.save}}</button>";
	
	let menu = "<div><p>Name:</p><input id=\\\"name\\\" value=\\\"" + _name + "\\\"/></div>"
	 + cameraOptions
	 + await getStream(data, name)
	 + await getPath(data, name)
	 + await getReason(data, name)
	 + "<div class=\\\"buttons\\\" id=\\\"buttons\\\">" + rem
	 + save
	 + "<button onclick=\\\"closeSettings();\\\">{{input.close}}</button></div>";
	
	let div = document.createElement("div");
	div.id = "sensor";
	div.classList = "sensor";
	div.innerHTML = "<div id=\\\"menu\\\" class=\\\"menu\\\">" + menu + "</div>";
	document.body.appendChild(div);
	
	let streamType = document.getElementById("streamType");
	if (streamType) streamType.addEventListener("change", async function(event) {
		let stream = document.getElementById("stream");
		if (stream) stream.innerHTML = await loadStream(streamType.value, name);
	});
	
	let pathType = document.getElementById("pathType");
	if (pathType) pathType.addEventListener("change", async function(event) {
		let path = document.getElementById("path");
		if (path) path.innerHTML = await loadPath(pathType.value, name);
	});
	
	let reasonType = document.getElementById("reasonType");
	if (reasonType) reasonType.addEventListener("change", async function(event) {
		let reason = document.getElementById("reason");
		if (reason) reason.innerHTML = await loadReason(reasonType.value, name);
	});
	
};
function removeCamera(name) {
	
	if (!name) return;
	
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status != 200) {
			let buttons = document.getElementById("buttons");
			inputError(buttons, this.response, true);
			return;
		};
		location.reload();
	};
	let sendData = {"cmd":"remove","name":name};
	xhr.send(JSON.stringify(sendData));
	
};
function power(name) {
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status != 200) {
			let buttons = document.getElementById("buttons");
			inputError(buttons, this.response, true);
			return;
		};
		location.reload();
	};
	let sendData = {"cmd":"power","name":name};
	xhr.send(JSON.stringify(sendData));
};
function glitch(name) {
	let xhr = new XMLHttpRequest();
	xhr.open("POST", "/cameras");
	xhr.onreadystatechange = function() {
		if (this.readyState != 4) return;
		if (this.status != 200) {
			let buttons = document.getElementById("buttons");
			inputError(buttons, this.response, true);
			return;
		};
		location.reload();
	};
	let sendData = {"cmd":"glitch","name":name};
	xhr.send(JSON.stringify(sendData));
};